### ISSUE TITLE
source-bundle: %{build-root} is not expanded to where the generated scripts run the builds

### ISSUE THREAD

**Issue by cs-shadow** [cs-shadow/buildstream#71]
In GitLab by @hashpling on Aug 21, 2017, 16:05

Given the following demo bst file:

```
kind: manual

depends:
- debian-stretch.bst

config:
  build-commands:
  - echo hello >world

  install-commands:
  - cp %{build-root}/world %{install-root}/world
```

The build and install commands work correctly in a normal sandboxed `bst build`, but the build script that `bst source-bundle` creates contains the following sequence of commands:

```
cp -a "$SRCDIR/hello-world" '/buildstream/build/hello-world'
cd '/buildstream/build/hello-world'

export PREFIX='/buildstream/install'

export HOME=/tmp LC_ALL=C LOGNAME=tomjon MAKEFLAGS=-j6 PATH=/usr/bin:/bin:/usr/sbin:/sbin SHELL=/bin/sh TERM=dumb TZ=UTC USER=tomjon USERNAME=tomjon V=1

# Build the module
(set -ex; echo hello >world
) || exit 1
(set -ex; cp /buildstream/build/world /buildstream/install/world
) || exit 1
(set -ex; :
) || exit 1
```

It seems that either the sources should be copied to `/buildstream/build` and not `/buildstream/build/<build-name>` (the whole directory is cleaned in any case), or `%{build-root}` should be expanded to the subdirectory that is actually used for the build.

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Aug 22, 2017, 15:13

> It seems that either the sources should be copied to /buildstream/build and not /buildstream/build/<build-name> (the whole directory is cleaned in any case), or %{build-root} should be expanded to the subdirectory that is actually used for the build.

Lets check with @tlater that this doesnt break anything else... but your assessment is correct.

**Comment by cs-shadow** [comment]
In GitLab by @hashpling on Aug 23, 2017, 15:39

mentioned in merge request !82

**Comment by cs-shadow** [comment]
In GitLab by @tlater on Aug 30, 2017, 10:27

!82 shouldn't break anything - the build directory is emptied before and after each build anyway. Changing what `{build-root}` expands to might be more future proof if parallel builds ever become desirable, but I think that's going a little far.

**Comment by cs-shadow** [comment]
In GitLab by @hashpling on Sep 5, 2017, 10:07

Yes, nothing is broken (that I know of!) in this approach. Going forward, though, I think there are good reasons for building in subdirectory of `/buildstream/build` and expanding `{bldroot}` appropriately, and not just from a `{source-bundle}` build.

One reason is debug information. Typically, the full path to source files is embedded in debug information. When the build has been performed in a sandbox, the source files aren't available at the recorded location at run / debug time. Often that's not a problem because most debuggers have some sort of path mapping feature, where you can specify path translations so that the debugger can find the source files at an alternative location.

There is a very positive potential upside to using a subdirectory based on the build element name. Having a unique and deterministic prefix to source paths means that it is simple to programmatically generate a source file mapping for a debugger and removes the possibility that different elements use the same path to refer to different files (or even the same file but built with different pre-processor options).

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Sep 5, 2017, 16:13

> Having a unique and deterministic prefix to source paths means that it is simple to programmatically generate a source file mapping for a debugger and removes the possibility that different elements use the same path to refer to different files (or even the same file but built with different pre-processor options).

Very good point I had previously missed.

I'm not sure how relevant this is to the `source-bundle` story but we should probably be doing some element name based substitutions on build paths in the regular case where things get built with buildstream.

Curiously, what are you using `source-bundle` for ?

To be frank, the `source-bundle` feature is a bit of a corner case; the only reason it exists is because without being able to export some build scripts which can be run on just a *basic* shell (like busybox), buildstream would have no viable cross bootstrapping story (so essentially you *have* to use `bst source-bundle` if you want to port to a new arch that you have to bootstrap yourself).

**Comment by cs-shadow** [comment]
In GitLab by @hashpling on Sep 5, 2017, 21:32

closed via commit b4f5f4ce20a486b57e96d15d8a2c2c6586e2ac5a

**Comment by cs-shadow** [comment]
In GitLab by @hashpling on Sep 9, 2017, 03:13

mentioned in commit b4f5f4ce20a486b57e96d15d8a2c2c6586e2ac5a
